### ISSUE TITLE
Return an immutable workspace from Solution.Workspace 

### ISSUE THREAD

**Issue by tmat** [dotnet/roslyn#35515]
A `Solution` instance is supposed to be (observably) immutable, but the class has a reference back to the `Workspace` the solution belongs to. This makes it easy for features to access a mutable `Workspace` state when they shouldn't and impossible to share solutions across workspaces.

Some of the information stored currently on `Workspace` should be moved to the `Solution` (e.g. some stateless workspace services).

We have a couple options:
1) Return `null`
2) Return an instance of an internal subclass of `Workspace`, which will provide the information we moved to the `Solution` instance via the existing `Workspace` APIs.

**Comment by jasonmalinowski** [comment]
I was thinking during the weekend: option 1 would break things, yes, but I'm I do wonder exactly what we might all break with option 2 as well, and whether we're best off just pulling the band-aid. The fake Workspace we'd have to make would act a bit strangely: what would CurrentSolution return? Does forking a Solution in this world create a new Workspace instance with a new CurrentSolution? What does that look like?

Looking at uses of Solution.Workspace, many of them are code refactorings. Maybe a "Workspace can return null, but in existing IDEs for the existing primary workspaces we'll still make return what it does today" is something going forward for short-term compat. But for newer Workspaces in other processes, we'll figure it out what it means.

If nothing else, I had the thought that when we're donig nullable annotations for the Roslyn API let's definitely mark Workspace as being nullable. Maybe we'll have to do 2 in some form for a compat shim, but long term let's plan on 1?

**Comment by CyrusNajmabadi** [comment]
I also gave this a bunch of thought, and it ranged from being terrified, to mostly being ok.  By and large i thought it would simply make sense to have .Workspace return a proxy **readonly** object that wrapped the true workspace (copying some data if necessary to appear immutable).  So basically, querying ops would work fine, but mutation ops would throw.  It would be equivalent to how you have APIs in dotnet that say "give me a readonly instance of this interface".  All the interface members are still there, but only acceptable ones do anything, and hte rest throw.   The copying would exist to prevent other mutation ops from being visible to this workspace.

This actually felt like it would be ok.

To answer your questions @jasonmalinowski:

> The fake Workspace we'd have to make would act a bit strangely: what would CurrentSolution return? Does forking a Solution in this world create a new Workspace instance with a new CurrentSolution? 

`solution.Workspace.CurrentSolution` would return `solution`.  If you forked, then `forkedSolution.Workspace.CurrentSolution` would return `forkedSolution`.  

That's my view on it at least.

--

Note: is there a critical issue we're trying to solve here? Or is this mostly just a hygiene concern?

**Comment by jasonmalinowski** [comment]
Yeah, what you're proposing is what @tmat was intending as the stub.

> Note: is there a critical issue we're trying to solve here? Or is this mostly just a hygiene concern?

Partly hygiene but this came up as one strange issue as we're looking at designing some other things.
